File: LR7/LR7/sierpinski.cpp
#include <SFML/Graphics.hpp>
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;
using namespace sf;

struct Point {
    float x, y;
    Point(float x = 0, float y = 0) : x(x), y(y) {}
};

class SierpinskiTriangle {
private:
    RenderWindow window;
    vector<VertexArray> triangles;
    int maxDepth;
    
    // Функция для получения средней точки между двумя точками
    Point getMidpoint(const Point& a, const Point& b) {
        return Point((a.x + b.x) / 2.0f, (a.y + b.y) / 2.0f);
    }
    
    // Рекурсивная функция для создания треугольника Серпинского
    void drawSierpinski(Point a, Point b, Point c, int depth) {
        if (depth == 0) {
            // Рисуем треугольник
            VertexArray triangle(Triangles, 3);
            triangle[0].position = Vector2f(a.x, a.y);
            triangle[1].position = Vector2f(b.x, b.y);
            triangle[2].position = Vector2f(c.x, c.y);
            
            // Устанавливаем цвет в зависимости от глубины
            Color color = Color::White;
            triangle[0].color = color;
            triangle[1].color = color;
            triangle[2].color = color;
            
            triangles.push_back(triangle);
        } else {
            // Находим средние точки сторон
            Point ab = getMidpoint(a, b);
            Point bc = getMidpoint(b, c);
            Point ca = getMidpoint(c, a);
            
            // Рекурсивно рисуем три меньших треугольника
            drawSierpinski(a, ab, ca, depth - 1);
            drawSierpinski(ab, b, bc, depth - 1);
            drawSierpinski(ca, bc, c, depth - 1);
        }
    }
    
public:
    SierpinskiTriangle(int width = 800, int height = 600, int depth = 5) 
        : window(VideoMode(width, height), "Треугольник Серпинского"), maxDepth(depth) {
        window.setFramerateLimit(60);
    }
    
    void generateTriangle() {
        triangles.clear();
        
        // Создаем начальный равносторонний треугольник
        float centerX = 400;
        float centerY = 300;
        float size = 250;
        
        // Вершины равностороннего треугольника
        Point a(centerX, centerY - size);
        Point b(centerX - size * cos(M_PI / 6), centerY + size * sin(M_PI / 6));
        Point c(centerX + size * cos(M_PI / 6), centerY + size * sin(M_PI / 6));
        
        drawSierpinski(a, b, c, maxDepth);
    }
    
    void handleEvents() {
        Event event;
        while (window.pollEvent(event)) {
            if (event.type == Event::Closed) {
                window.close();
            }
            
            if (event.type == Event::KeyPressed) {
                if (event.key.code == Keyboard::Up && maxDepth < 8) {
                    maxDepth++;
                    generateTriangle();
                    cout << "Глубина увеличена до: " << maxDepth << endl;
                }
                if (event.key.code == Keyboard::Down && maxDepth > 0) {
                    maxDepth--;
                    generateTriangle();
                    cout << "Глубина уменьшена до: " << maxDepth << endl;
                }
                if (event.key.code == Keyboard::R) {
                    generateTriangle();
                    cout << "Треугольник перерисован" << endl;
                }
            }
        }
    }
    
    void render() {
        window.clear(Color::Black);
        
        // Рисуем все треугольники
        for (const auto& triangle : triangles) {
            window.draw(triangle);
        }
        
        // Создаем текст с информацией
        Font font;
        window.display();
    }
    
    void run() {
        cout << "Треугольник Серпинского" << endl;
        cout << "Управление:" << endl;
        cout << "Стрелка вверх - увеличить глубину рекурсии" << endl;
        cout << "Стрелка вниз - уменьшить глубину рекурсии" << endl;
        cout << "R - перерисовать треугольник" << endl;
        cout << "Текущая глубина: " << maxDepth << endl;
        
        generateTriangle();
        
        while (window.isOpen()) {
            handleEvents();
            render();
        }
    }
};

int main() {
    try {
        SierpinskiTriangle sierpinski(800, 600, 3);
        sierpinski.run();
    }
    catch (const exception& e) {
        cout << "Ошибка: " << e.what() << endl;
        return -1;
    }
    
    return 0;
}

File: LR7/LR7/sierpinski.py
import pygame
import math
import sys

class Point:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

class SierpinskiTriangle:
    def __init__(self, width=800, height=600, depth=5):
        pygame.init()
        self.width = width
        self.height = height
        self.screen = pygame.display.set_mode((width, height))
        pygame.display.set_caption("Треугольник Серпинского")
        self.clock = pygame.time.Clock()
        
        self.triangles = []
        self.max_depth = depth
        
        # Цвета
        self.BLACK = (0, 0, 0)
        self.WHITE = (255, 255, 255)
        
    def get_midpoint(self, a, b):
        """Функция для получения средней точки между двумя точками"""
        return Point((a.x + b.x) / 2.0, (a.y + b.y) / 2.0)
    
    def draw_sierpinski(self, a, b, c, depth):
        """Рекурсивная функция для создания треугольника Серпинского"""
        if depth == 0:
            # Добавляем треугольник в список для отрисовки
            triangle = [(a.x, a.y), (b.x, b.y), (c.x, c.y)]
            self.triangles.append(triangle)
        else:
            # Находим средние точки сторон
            ab = self.get_midpoint(a, b)
            bc = self.get_midpoint(b, c)
            ca = self.get_midpoint(c, a)
            
            # Рекурсивно рисуем три меньших треугольника
            self.draw_sierpinski(a, ab, ca, depth - 1)
            self.draw_sierpinski(ab, b, bc, depth - 1)
            self.draw_sierpinski(ca, bc, c, depth - 1)
    
    def generate_triangle(self):
        """Генерация треугольника Серпинского"""
        self.triangles.clear()
        
        # Создаем начальный равносторонний треугольник
        center_x = 400
        center_y = 300
        size = 250
        
        # Вершины равностороннего треугольника
        a = Point(center_x, center_y - size)
        b = Point(center_x - size * math.cos(math.pi / 6), 
                 center_y + size * math.sin(math.pi / 6))
        c = Point(center_x + size * math.cos(math.pi / 6), 
                 center_y + size * math.sin(math.pi / 6))
        
        self.draw_sierpinski(a, b, c, self.max_depth)
    
    def handle_events(self):
        """Обработка событий"""
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return False
            
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_UP and self.max_depth < 8:
                    self.max_depth += 1
                    self.generate_triangle()
                    print(f"Глубина увеличена до: {self.max_depth}")
                elif event.key == pygame.K_DOWN and self.max_depth > 0:
                    self.max_depth -= 1
                    self.generate_triangle()
                    print(f"Глубина уменьшена до: {self.max_depth}")
                elif event.key == pygame.K_r:
                    self.generate_triangle()
                    print("Треугольник перерисован")
        
        return True
    
    def render(self):
        """Отрисовка"""
        self.screen.fill(self.BLACK)
        
        # Рисуем все треугольники
        for triangle in self.triangles:
            pygame.draw.polygon(self.screen, self.WHITE, triangle)
        
        pygame.display.flip()
    
    def run(self):
        """Главный цикл программы"""
        print("Треугольник Серпинского")
        print("Управление:")
        print("Стрелка вверх - увеличить глубину рекурсии")
        print("Стрелка вниз - уменьшить глубину рекурсии")
        print("R - перерисовать треугольник")
        print(f"Текущая глубина: {self.max_depth}")
        
        self.generate_triangle()
        
        running = True
        while running:
            running = self.handle_events()
            self.render()
            self.clock.tick(60)  # 60 FPS
        
        pygame.quit()
        sys.exit()

def main():
    try:
        sierpinski = SierpinskiTriangle(800, 600, 3)
        sierpinski.run()
    except Exception as e:
        print(f"Ошибка: {e}")
        return -1
    
    return 0

if __name__ == "__main__":
    main()